### PR TITLE
Correctly notify the creator of article, if it began as a redirect

### PR DESCRIPTION
**Pull request by siddharthvp**
I have tweaked Morebits.wiki.page.prototype.lookupCreation so that an option is now available which if set, the function would return the username/timestamp of the creation of the first non-redirect version of the article. 

I have then put this into use for AFD and PROD notifications, since in these cases we unambiguously want to notify the real article creator rather than the redirect creator. We should perhaps also try to incorporate this somehow to CSD, but I guess that it's a bit trickier. 

I guess this is long-requested feature, and should be very helpful to the project as now people who create articles from redirects would not remain in the dark when their work is sent to the gallows. Also, this should help us crush the upcoming competition from PageCuration!

### CODE REVIEW

**Comment by siddharthvp** [comment]
Code workflow: A regex test for `/^\s*#redirect/i' is used to check if a version is a redirect. This does not break internationalization, as the same syntax is used for redirects even in non-English language wikis. The tags ("Removing redirect"/"Redirect becoming article") are _not_ being used since they didn't aren't there in early revisions, nor on other wikis. 

At first, only the first revision is fetched, and that too only the the top section (to save bandwidth). If it turns out that it was a redirect and the `lookupNonRedirectCreator` option is enabled, the 50 earliest revisions are now fetched (again only rvsection=0) and we iterate to find the first non-redirect version.

File: modules/twinklexfd.js [hunk 1]
```diff
@@ -716,6 +716,7 @@ Twinkle.xfd.callbacks = {
 			if (params.usertalk) {
 				var thispage = new Morebits.wiki.page(mw.config.get('wgPageName'));
 				thispage.setCallbackParameters(params);
+				thispage.setLookupNonRedirectCreator(true);
```
Comment (Amorymeltzer) [inline modules/twinklexfd.js]:
```suggestion
				thispage.setLookupNonRedirectCreator(true); // Look for first non-redirect author
```

File: modules/twinkleprod.js [hunk 2]
```diff
@@ -204,6 +204,7 @@ Twinkle.prod.callbacks = {
 		var ts = new Morebits.wiki.page(mw.config.get('wgPageName'));
 		ts.setFollowRedirect(true);  // for NPP, and also because redirects are ineligible for PROD
 		ts.setCallbackParameters(params);
+		ts.setLookupNonRedirectCreator(true);
```
Comment (Amorymeltzer) [inline modules/twinkleprod.js]:
```suggestion
		ts.setLookupNonRedirectCreator(true); // Look for first non-redirect author
```

File: morebits.js [hunk 3]
```diff
@@ -2708,18 +2717,53 @@ Morebits.wiki.page = function(pageName, currentAction) {
 			return; // abort
 		}
 
-		ctx.creator = $(xml).find('rev').attr('user');
+		if (!ctx.lookupNonRedirectCreator || !/^\s*#redirect/i.test($(xml).find('rev').text())) {
```
[the remark this answers is not in the document]
Comment (siddharthvp) [reply morebits.js]: > having the typical behavior first is nice

That's what I thought.

File: morebits.js [hunk 4]
```diff
@@ -2708,18 +2717,53 @@ Morebits.wiki.page = function(pageName, currentAction) {
...
+
+		$(xml).find('rev').each(function(_, rev) {
+			if (!/^\s*#redirect/i.test(rev.textContent)) {
+				ctx.creator = rev.getAttribute('user');
+				ctx.timestamp = rev.getAttribute('timestamp');
+				return false; // break
+			}
+		});
+
 		if (!ctx.creator) {
```
[the remark this answers is not in the document]
Comment (siddharthvp) [reply morebits.js]: Good idea! Will do.

File: morebits.js [hunk 5]
```diff
@@ -2041,6 +2044,21 @@ Morebits.wiki.page = function(pageName, currentAction) {
...
+	/**
+	 * @param {boolean} flag - if set true, the author and timestamp of the first non-redirect
+	 * version of the page is retrieved.
+	 *
+	 * Warning:
+	 * 1. If there are no revisions among the first 50 that are non-redirects, or if there are
+	 *    less 50 revisions and all are redirects, the original creation is retrived.
+	 * 2. Revisions that the user is not privileged to access (revdeled/suppressed) will be treated
+	 *    as non-redirects.
+	 */
```
Comment (siddharthvp) [inline morebits.js]: This makes sense as revdels in main namespace is usually for copyright violations, and only non-redirect revisions would have been revdeled as such.

File: morebits.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (siddharthvp) [reply morebits.js]: done.

File: morebits.js (hunk 5 above)
Comment (Amorymeltzer) [reply morebits.js]: Doesn't really make a difference either way: not much else we can do, and it should be pretty rare regardless.  The reality is that early vandalism is going to be an issue no matter what, as articles that were initially redirects but had someone vandalize that redirect before becoming an article will still notify the wrong person.  In the end, though, this is basically taking a simple process and making it somewhat smarter, but that doesn't mean we need to account for every single contingency.

**Pull request merged**